pkg/controller/dns/letsencrypt.go

```diff
@@ -0,0 +1,349 @@
+	}
+
+	if u.URL == "" {
+		return fmt.Errorf("not registering LE User: missing URL")
+	}
+
+	if u.Key == nil {
+		key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		if err != nil {
+			return err
+		}
+		u.Key = key
+	}
+
+	conf := lego.NewConfig(u)
+
+	conf.CADirURL = u.URL
+
+	client, err := lego.NewClient(conf)
+	if err != nil {
+		return err
+	}
+
+	reg, err := client.Registration.Register(registration.RegisterOptions{TermsOfServiceAgreed: true})
+	if err != nil {
+		return err
+	}
+	u.Registration = reg
+
+	logrus.Infof("registered LE User: %s", u.Name)
+
+	return nil
+
+}
+
+func (u *LEUser) ToSecret() (*corev1.Secret, error) {
+	if u.Registration == nil {
+		return nil, fmt.Errorf("not saving LE User: missing registration")
+	}
+	x509Encoded, _ := x509.MarshalECPrivateKey(u.Key.(*ecdsa.PrivateKey))
+	pemEncoded := pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: x509Encoded})
+
+	reg, err := json.Marshal(u.Registration)
+	if err != nil {
+		return nil, err
+	}
+
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      system.LESecretName,
+			Namespace: system.Namespace,
+			Labels: map[string]string{
+				labels.AcornManaged: "true",
+			},
+		},
+		Data: map[string][]byte{
+			"name":         []byte(u.Name),
+			"email":        []byte(u.Email),
+			"privateKey":   pemEncoded,
+			"registration": reg,
+			"url":          []byte(u.URL),
+		},
+	}, nil
+}
+
+func (u *LEUser) GenerateWildcardCert(dnsendpoint, domain, token string) (*certificate.Resource, error) {
+	if u.Registration == nil {
+		return nil, fmt.Errorf("not generating LE cert: missing registration")
+	}
+
+	if domain == "" || token == "" {
+		return nil, fmt.Errorf("not generating LE cert: missing domain or token")
+	}
+
+	conf := lego.NewConfig(u)
+
+	conf.CADirURL = u.URL
+
+	client, err := lego.NewClient(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	dnsProvider := NewDNSProvider(dnsendpoint, domain, token)
+
+	if err := client.Challenge.SetDNS01Provider(dnsProvider); err != nil {
+		return nil, err
+	}
+
+	request := certificate.ObtainRequest{
+		Domains: []string{fmt.Sprintf("*.%s", strings.TrimPrefix(domain, "."))},
+		Bundle:  true,
+	}
+
+	return client.Certificate.Obtain(request)
+}
+
+func matchLeURLToEnv(url string) string {
+	if url == LetsEncryptURLStaging {
+		return "staging"
+	} else if url == LetsEncryptURLProduction {
+		return "production"
+	} else {
+		return "disabled"
+	}
+}
+
+func (u *LEUser) toHash() string {
+	toHash := []byte(fmt.Sprintf("%s-%s-%s", u.Name, matchLeURLToEnv(u.URL), u.Email))
+	dig := sha1.New()
+	dig.Write([]byte(toHash))
+	return hex.EncodeToString(dig.Sum(nil))
+}
+
+func EnsureLEUser(ctx context.Context, cfg *apiv1.Config, client kclient.Client, domain string) (*LEUser, error) {
+
+	/*
+	 * Construct new LE User
+	 */
+	targetUsername := strings.TrimPrefix(domain, ".") // leading dot can be an issue
+	email := fmt.Sprintf("%s@on-acorn.io", targetUsername)
```

Comment:
lets just hard code the staging email to `staging-certs@acorn.io`. I'm going to create that as a real email address so they dont bounce.